miniapps/navier/ortho_solver.cpp

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) 2010-2020, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#include "ortho_solver.hpp"
+
+using namespace mfem;
+using namespace navier;
+
+OrthoSolver::OrthoSolver() : Solver(0, true) {}
+
+void OrthoSolver::SetOperator(const Operator &op)
+{
+   oper = &op;
+}
+
+void OrthoSolver::Mult(const Vector &b, Vector &x) const
+{
+   // Orthoganlize input.
+   Orthoganalize(b, b_ortho);
+
+   // Apply operator.
+   oper->Mult(b_ortho, x);
+
+   // Orthoganlize output.
+   Orthoganalize(x, x);
+}
+
+void OrthoSolver::Orthoganalize(const Vector &v, Vector &v_ortho) const
+{
+   double loc_sum = v.Sum();
+   double global_sum = 0.0;
+   int loc_size = v.Size();
+   int global_size = 0;
+
+   MPI_Allreduce(&loc_sum, &global_sum, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);
+   MPI_Allreduce(&loc_size, &global_size, 1, MPI_INT, MPI_SUM, MPI_COMM_WORLD);
+
+   double ratio = global_sum / static_cast<double>(global_size);
+   v_ortho.SetSize(v.Size());
+   for (int i = 0; i < v_ortho.Size(); ++i)
+   {
+      v_ortho(i) = v(i) - ratio;
+   }
+}
```

Comment:
This is essentially the same as [NavierSolver::Orthogonalize](https://github.com/mfem/mfem/blob/22e788765998ee8b52e2f0645c2f0499f9741dab/miniapps/navier/navier_solver.cpp#L574). Can we avoid redundancy here? Maybe we could add this to the Vector class instead?

Reply:
It is a shared utility function. But actually the only one. So for simplicity I just copied method.

Reply:
Okay. Btw I noticed use of "Orthog**a**nalize" in `OrthoSolver` and "Orthog**o**nalize" in `NavierSolver`. We should make it consistent across both classes.